File: 213-house-robber-ii/213-house-robber-ii.java
class Solution {
    private static int houseRob(int[] nums, int start, int end) {
        int prev2 = 0, prev = nums[start];
        for(int idx = start + 1; idx <= end; ++idx) {
            int current = Math.max(nums[idx] + prev2, prev);
            prev2 = prev;
            prev = current;
        }
        return prev;
    }
    public int rob(int[] nums) {
        int n = nums.length;
        if(n == 1) return nums[0];
        return Math.max(houseRob(nums, 0, n - 2), houseRob(nums, 1, n - 1));
    }
}

File: 213-house-robber-ii/213-house-robber-ii.py
class Solution:
    def maxRob(self, nums: List[int], start: int, end: int) -> int:
        prev2, prev = 0, nums[start]
        for idx in range(start + 1, end + 1):
            current = max(nums[idx] + prev2, prev)
            prev2 = prev
            prev = current
        return prev
        
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1: return nums[0]
        return max(self.maxRob(nums, 0, n - 2), self.maxRob(nums, 1, n - 1))